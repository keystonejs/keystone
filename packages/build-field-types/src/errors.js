// @flow

export class FatalError extends Error {
  item: { +name: string };
  constructor(message: string, item: { +name: string }) {
    super(message);
    this.item = item;
  }
}

export class ValidationError extends Error {}

export class MissingDependency extends Error {}

export class FixableError extends FatalError {}
