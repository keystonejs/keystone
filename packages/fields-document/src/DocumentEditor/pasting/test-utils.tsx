/** @jsxRuntime classic */
/** @jsx jsx */
import { makeEditor, jsx } from '../tests/utils';
import { MyDataTransfer } from './data-transfer';

export function htmlToEditor(html: string) {
  let editor = makeEditor(
    <editor>
      <paragraph>
        <text>
          <cursor />
        </text>
      </paragraph>
    </editor>
  );
  const data = new MyDataTransfer();
  data.setData('text/html', html);
  editor.insertData(data);
  return editor;
}
