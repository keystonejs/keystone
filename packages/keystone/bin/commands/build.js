const chalk = require('chalk');
const path = require('path');
const fs = require('fs-extra');
const { getEntryFileFullPath } = require('../utils');
const { DEFAULT_ENTRY, DEFAULT_DIST_DIR } = require('../../constants');

module.exports = {
  // prettier-ignore
  spec: {
    '--out':   String,
    '-o':      '--out',
    '--entry': String,
  },
  help: ({ exeName }) => `
    Usage
      $ ${exeName} build --out=dist

    Options
      --out, -o   Directory to save build [dist]
      --entry     Entry file exporting keystone instance [${DEFAULT_ENTRY}]
  `,
  exec: async (args, { exeName, _cwd = process.cwd() } = {}, spinner) => {
    process.env.NODE_ENV = 'production';

    spinner.text = 'Validating project entry file';
    let entryFile = await getEntryFileFullPath(args, { exeName, _cwd });
    spinner.succeed(`Validated project entry file ./${path.relative(_cwd, entryFile)}`);

    spinner.start('Initialising Keystone instance');
    let { keystone, apps, distDir = DEFAULT_DIST_DIR } = require(entryFile);
    spinner.succeed('Initialised Keystone instance');

    if (args['--out']) {
      distDir = args['--out'];
    }
    let resolvedDistDir = path.resolve(_cwd, distDir);
    spinner.start(`Exporting Keystone build to ./${path.relative(_cwd, resolvedDistDir)}`);

    await fs.remove(resolvedDistDir);

    if (apps) {
      await Promise.all(
        apps.filter(app => app.build).map(app => app.build({ distDir: resolvedDistDir, keystone }))
      );

      spinner.succeed(
        chalk.green.bold(`Exported Keystone build to ./${path.relative(_cwd, resolvedDistDir)}`)
      );
    } else {
      spinner.info('Nothing to build.');
      spinner.info(
        `To create an Admin UI build, make sure you export 'admin' from ./${path.relative(
          _cwd,
          entryFile
        )}`
      );
    }

    // It's possible the developer's entry file has an accidentally long-lived
    // process (for example, using 'connect-mongodb-session').
    // To avoid that long-lived process from making the build command hang, we
    // signal that we wish the process to end here.
    return { endProcessWithSuccess: true };
  },
};
