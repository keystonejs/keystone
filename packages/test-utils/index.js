const {
  setupServer,
  multiAdapterRunners,
  graphqlRequest,
  authedGraphqlRequest,
  networkedGraphqlRequest,
  matchFilter,
} = require('./lib/test-utils');

module.exports = {
  setupServer,
  multiAdapterRunners,
  graphqlRequest,
  authedGraphqlRequest,
  networkedGraphqlRequest,
  matchFilter,
};
