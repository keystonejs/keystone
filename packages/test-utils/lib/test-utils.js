const express = require('express');
const supertest = require('supertest-light');
const MongoDBMemoryServer = require('mongodb-memory-server-core').default;
const pFinally = require('p-finally');
const url = require('url');
const { Keystone } = require('@keystonejs/keystone');
const { GraphQLApp } = require('@keystonejs/app-graphql');
const { KnexAdapter } = require('@keystonejs/adapter-knex');
const { MongooseAdapter } = require('@keystonejs/adapter-mongoose');

async function setupServer({
  adapterName,
  schemaName = 'testing',
  schemaNames = ['testing'],
  createLists = () => {},
  keystoneOptions,
  graphqlOptions = {},
}) {
  const Adapter = { mongoose: MongooseAdapter, knex: KnexAdapter }[adapterName];

  const argGenerator = {
    mongoose: getMongoMemoryServerConfig,
    knex: () => ({
      dropDatabase: true,
      knexOptions: { connection: process.env.KNEX_URI || 'postgres://localhost/keystone' },
    }),
  }[adapterName];

  const keystone = new Keystone({
    adapter: new Adapter(await argGenerator()),
    defaultAccess: { list: true, field: true },
    schemaNames,
    cookieSecret: 'secretForTesting',
    ...keystoneOptions,
  });

  createLists(keystone);

  const apps = [
    new GraphQLApp({
      schemaName,
      apiPath: '/admin/api',
      graphiqlPath: '/admin/graphiql',
      apollo: {
        tracing: true,
        cacheControl: {
          defaultMaxAge: 3600,
        },
      },
      ...graphqlOptions,
    }),
  ];

  const { middlewares } = await keystone.prepare({ dev: true, apps });

  const app = express();
  app.use(middlewares);

  return { keystone, app };
}

function graphqlRequest({ keystone, query, variables }) {
  return keystone.executeGraphQL({
    context: keystone.createContext({ schemaName: 'testing', skipAccessControl: true }),
    query,
    variables,
  });
}

// This is much like graphqlRequest except we don't skip access control checks!
function authedGraphqlRequest({ keystone, query, variables }) {
  return keystone.executeGraphQL({
    context: keystone.createContext({
      schemaName: 'testing',
      authentication: {},
      skipAccessControl: false,
    }),
    query,
    variables,
  });
}

function networkedGraphqlRequest({
  app,
  query,
  variables = undefined,
  headers = {},
  expectedStatusCode = 200,
  operationName,
}) {
  const request = supertest(app).set('Accept', 'application/json');

  Object.entries(headers).forEach(([key, value]) => request.set(key, value));

  return request
    .post('/admin/api', { query, variables, operationName })
    .then(res => {
      expect(res.statusCode).toBe(expectedStatusCode);
      return {
        ...JSON.parse(res.text),
        res,
      };
    })
    .catch(error => ({
      errors: [error],
    }));
}

// One instance per node.js thread which cleans itself up when the main process
// exits
let mongoServer;
let mongoServerReferences = 0;

async function getMongoMemoryServerConfig() {
  mongoServer = mongoServer || new MongoDBMemoryServer();
  mongoServerReferences++;
  // Passing `true` here generates a new, random DB name for us
  const mongoUri = await mongoServer.getConnectionString(true);
  // In theory the dbName can contain query params so lets parse it then extract the db name
  const dbName = url
    .parse(mongoUri)
    .pathname.split('/')
    .pop();

  return { mongoUri, dbName };
}

function teardownMongoMemoryServer() {
  mongoServerReferences--;
  if (mongoServerReferences < 0) {
    mongoServerReferences = 0;
  }

  if (mongoServerReferences > 0) {
    return Promise.resolve();
  }

  if (!mongoServer) {
    return Promise.resolve();
  }
  const stopping = mongoServer.stop();
  mongoServer = null;
  return stopping;
}

function getCreate(keystone) {
  return (list, item) => keystone.getListByKey(list).adapter.create(item);
}

function getFindById(keystone) {
  return (list, item) => keystone.getListByKey(list).adapter.findById(item);
}

function getFindOne(keystone) {
  return (list, item) => keystone.getListByKey(list).adapter.findOne(item);
}

function getUpdate(keystone) {
  return (list, id, data) => keystone.getListByKey(list).adapter.update(id, data);
}

function getDelete(keystone) {
  return (list, id) => keystone.getListByKey(list).adapter.delete(id);
}

function _keystoneRunner(adapterName, tearDownFunction) {
  return function(setupKeystoneFn, testFn) {
    return async function() {
      if (!testFn) {
        // If a testFn is not defined then we just need
        // to excute setup and tear down in isolation.
        try {
          await setupKeystoneFn(adapterName);
        } catch (error) {
          await tearDownFunction();
          throw error;
        }
        return;
      }
      const setup = await setupKeystoneFn(adapterName);
      const { keystone } = setup;

      await keystone.connect();

      return pFinally(
        testFn({
          ...setup,
          create: getCreate(keystone),
          findById: getFindById(keystone),
          findOne: getFindOne(keystone),
          update: getUpdate(keystone),
          delete: getDelete(keystone),
        }),
        () => keystone.disconnect().then(tearDownFunction)
      );
    };
  };
}

function _before(adapterName) {
  return async function(setupKeystone) {
    const { keystone, app } = await setupKeystone(adapterName);
    await keystone.connect();
    return { keystone, app };
  };
}

function _after(tearDownFunction) {
  return async function(keystone) {
    await keystone.disconnect();
    await tearDownFunction();
  };
}

function multiAdapterRunners(only) {
  return [
    {
      runner: _keystoneRunner('mongoose', teardownMongoMemoryServer),
      adapterName: 'mongoose',
      before: _before('mongoose'),
      after: _after(teardownMongoMemoryServer),
    },
    {
      runner: _keystoneRunner('knex', () => {}),
      adapterName: 'knex',
      before: _before('knex'),
      after: _after(() => {}),
    },
  ].filter(a => typeof only === 'undefined' || a.adapterName === only);
}

const sorted = (arr, keyFn) => {
  arr = [...arr];
  arr.sort((a, b) => {
    a = keyFn(a);
    b = keyFn(b);
    if (a < b) {
      return -1;
    }
    if (a > b) {
      return 1;
    }
    return 0;
  });
  return arr;
};

const matchFilter = ({ keystone, queryArgs, fieldSelection, expected, sortKey }) => {
  return graphqlRequest({
    keystone,
    query: `query {
      allTests${queryArgs ? `(${queryArgs})` : ''} { ${fieldSelection} }
    }`,
  }).then(({ data }) => {
    const value = sortKey ? sorted(data.allTests || [], i => i[sortKey]) : data.allTests;
    expect(value).toEqual(expected);
  });
};

class MockFieldImplementation {
  constructor() {
    this.access = {
      public: {
        create: false,
        read: true,
        update: false,
        delete: false,
      },
    };
    this.config = {};
    this.hooks = {};
  }
  getAdminMeta() {
    return { path: 'id' };
  }
  gqlOutputFields() {
    return ['id: ID'];
  }
  gqlQueryInputFields() {
    return ['id: ID'];
  }
  get gqlUpdateInputFields() {
    return ['id: ID'];
  }
  get gqlCreateInputFields() {
    return ['id: ID'];
  }
  getGqlAuxTypes() {
    return [];
  }
  getGqlAuxQueries() {
    return [];
  }
  getGqlAuxMutations() {
    return [];
  }
  gqlOutputFieldResolvers() {
    return {};
  }
  gqlAuxQueryResolvers() {
    return {};
  }
  gqlAuxMutationResolvers() {
    return {};
  }
  gqlAuxFieldResolvers() {
    return {};
  }
  extendAdminViews(views) {
    return views;
  }
  getDefaultValue() {
    return;
  }
  async resolveInput({ resolvedData }) {
    return resolvedData.id;
  }
  async validateInput() {}
  async beforeChange() {}
  async afterChange() {}
  async beforeDelete() {}
  async validateDelete() {}
  async afterDelete() {}
}
class MockFieldAdapter {}

const MockIdType = {
  implementation: MockFieldImplementation,
  views: {},
  adapters: { mock: MockFieldAdapter },
};

class MockListAdapter {
  name = 'mock';
  constructor(parentAdapter) {
    this.parentAdapter = parentAdapter;
    this.index = 3;
    this.items = {
      0: { name: 'a', email: 'a@example.com', index: 0 },
      1: { name: 'b', email: 'b@example.com', index: 1 },
      2: { name: 'c', email: 'c@example.com', index: 2 },
    };
  }
  newFieldAdapter = () => new MockFieldAdapter();
  create = async item => {
    this.items[this.index] = {
      ...item,
      index: this.index,
    };
    this.index += 1;
    return this.items[this.index - 1];
  };
  findById = id => this.items[id];
  delete = async id => {
    this.items[id] = undefined;
  };
  itemsQuery = async ({ where: { id_in: ids, id, id_not_in } }, { meta = false } = {}) => {
    if (meta) {
      return {
        count: (id !== undefined
          ? [this.items[id]]
          : ids.filter(i => !id_not_in || !id_not_in.includes(i)).map(i => this.items[i])
        ).length,
      };
    } else {
      return id !== undefined
        ? [this.items[id]]
        : ids.filter(i => !id_not_in || !id_not_in.includes(i)).map(i => this.items[i]);
    }
  };
  itemsQueryMeta = async args => this.itemsQuery(args, { meta: true });
  update = (id, item) => {
    this.items[id] = { ...this.items[id], ...item };
    return this.items[id];
  };
}

class MockAdapter {
  name = 'mock';
  newListAdapter = () => new MockListAdapter(this);
  getDefaultPrimaryKeyConfig = () => ({ type: MockIdType });
}

module.exports = {
  setupServer,
  multiAdapterRunners,
  graphqlRequest,
  authedGraphqlRequest,
  networkedGraphqlRequest,
  matchFilter,
  MockAdapter,
  MockListAdapter,
  MockIdType,
};
