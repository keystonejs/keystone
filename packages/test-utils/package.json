{
  "name": "@keystonejs/test-utils",
  "description": "Common utilities used while testing @keystonejs/* packages.",
  "version": "6.1.3",
  "author": "The KeystoneJS Development Team",
  "license": "MIT",
  "engines": {
    "node": ">=10.0.0"
  },
  "dependencies": {
    "@keystonejs/adapter-knex": "^10.0.2",
    "@keystonejs/adapter-mongoose": "^8.1.2",
    "@keystonejs/app-graphql": "^5.1.8",
    "@keystonejs/keystone": "^10.0.0",
    "express": "^4.17.1",
    "mongodb-memory-server-core": "^6.5.2",
    "p-finally": "^2.0.1",
    "supertest-light": "^1.0.3"
  },
  "repository": "https://github.com/keystonejs/keystone/tree/master/packages/test-utils"
}
