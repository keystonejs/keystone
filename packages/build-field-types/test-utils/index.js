// @flow
/* eslint-disable import/no-extraneous-dependencies */
import path from 'path';
import * as fs from 'fs-extra';
import globby from 'globby';
import fixturez from 'fixturez';
import spawn from 'spawndamnit';

let f = fixturez(__dirname);

require('chalk').enabled = false;
// $FlowFixMe
console.error = jest.fn();
// $FlowFixMe
console.log = jest.fn();

export let logMock = {
  log: ((console.log: any): JestMockFn<any, void>),
  error: ((console.error: any): JestMockFn<any, void>),
};

afterEach(() => {
  logMock.log.mockReset();
  logMock.error.mockReset();
});

import init from '../src/init';
import { confirms } from '../src/messages';

export async function initBasic(directory: string) {
  confirms.writeMainModuleFields.mockReturnValue(true);
  await init(directory);
  confirms.writeMainModuleFields.mockReset();
}

function getPkgPath(tmpPath: string) {
  return path.join(tmpPath, 'package.json');
}

export async function getPkg(filepath: string): Object {
  return JSON.parse(await fs.readFile(path.join(filepath, 'package.json'), 'utf-8'));
}

export async function modifyPkg(tmpPath: string, cb: Object => mixed) {
  let json = await getPkg(tmpPath);
  await cb(json);

  let pkgPath = getPkgPath(tmpPath);
  await fs.writeFile(pkgPath, JSON.stringify(json, null, 2));
}

export let createPackageCheckTestCreator = (doResult: string => Promise<void>) => {
  let createTestCreator = testFn => async (
    testName: string,
    entrypoints: { [key: string]: Object },
    cb: (doThing: () => Promise<{ [key: string]: Object }>) => Promise<void>
  ) => {
    testFn(testName, async () => {
      let tmpPath = f.copy('template-simple-package');
      let things = Object.keys(entrypoints);
      await Promise.all(
        things.map(async entrypointPath => {
          let content = entrypoints[entrypointPath];
          let filepath = path.join(tmpPath, entrypointPath, 'package.json');
          await fs.ensureFile(filepath);
          await fs.writeFile(filepath, JSON.stringify(content, null, 2));
        })
      );

      await cb(async () => {
        await doResult(tmpPath);

        let newThings = {};

        await Promise.all(
          things.map(async entrypointPath => {
            newThings[entrypointPath] = JSON.parse(
              await fs.readFile(path.join(tmpPath, entrypointPath, 'package.json'), 'utf8')
            );
          })
        );
        return newThings;
      });
    });
  };
  let testFn = createTestCreator(test);
  testFn.only = createTestCreator(test.only);
  testFn.skip = createTestCreator(test.skip);
  return testFn;
};

export async function snapshotDistFiles(tmpPath: string) {
  let distPath = path.join(tmpPath, 'dist');
  let distFiles;
  try {
    distFiles = await fs.readdir(distPath);
  } catch (err) {
    if (err.code === 'ENOENT') {
      throw new Error(distPath + ' does not exist');
    }
    throw err;
  }

  await Promise.all(
    distFiles.map(async x => {
      expect(await fs.readFile(path.join(distPath, x), 'utf-8')).toMatchSnapshot(x);
    })
  );
}

export let stripHashes = (chunkName: string) => {
  let transformer = (pathname: string) => {
    return pathname.replace(
      new RegExp(`(${chunkName}-[^\\.]+|dist\\/[^\\.]+\\/package.json)`, 'g'),
      () => {
        return `chunk-this-is-not-the-real-hash`;
      }
    );
  };
  return {
    transformPath: transformer,
    transformContent: (content: string) => {
      return content.replace(new RegExp(`${chunkName}-[^\\.]+`, 'g'), () => {
        return 'chunk-some-hash';
      });
    },
  };
};

export async function snapshotDirectory(
  tmpPath: string,
  {
    files = 'js',
    transformPath = x => x,
    transformContent = x => x,
  }: {
    files?: 'all' | 'js',
    transformPath?: (path: string, contents: string) => string,
    transformContent?: (content: string) => string,
  } = {}
) {
  let paths = await globby(
    [`**/${files === 'js' ? '*.js' : '*'}`, '!node_modules/**', '!yarn.lock'],
    {
      cwd: tmpPath,
    }
  );

  await Promise.all(
    paths.map(async x => {
      let content = transformContent(await fs.readFile(path.join(tmpPath, x), 'utf-8'));
      if (x.endsWith('.json')) {
        content = JSON.parse(content);
      }
      expect(content).toMatchSnapshot(transformPath(x, content));
    })
  );
}

export async function install(tmpPath: string) {
  await spawn('yarn', ['install'], { cwd: tmpPath });
}
