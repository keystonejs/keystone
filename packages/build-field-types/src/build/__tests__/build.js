// @flow
import build from '../';
import fixturez from 'fixturez';
import path from 'path';
import { initBasic, snapshotDistFiles, install } from '../../../test-utils';
import { FatalError } from '../../errors';

const f = fixturez(__dirname);

jest.mock('../../prompt');

jest.mock('install-packages');

let unsafeRequire = require;

test('monorepo', async () => {
  let tmpPath = f.copy('monorepo');
  await initBasic(tmpPath);
  await install(tmpPath);
  await build(tmpPath);
  let counter = 1;
  for (let pkg of ['package-one', 'package-two']) {
    let pkgPath = path.join(tmpPath, 'packages', pkg);
    await snapshotDistFiles(pkgPath);

    expect(unsafeRequire(pkgPath).default).toBe(counter++);
  }
});

// TODO: make it faster so this isn't required
jest.setTimeout(20000);

test('@babel/runtime installed', async () => {
  let tmpPath = f.copy('babel-runtime-installed');

  await install(tmpPath);

  await build(tmpPath);

  await snapshotDistFiles(tmpPath);
});

test('monorepo single package', async () => {
  let tmpPath = f.copy('monorepo-single-package');
  await initBasic(tmpPath);
  await install(tmpPath);

  await build(tmpPath);
  let pkgPath = path.join(tmpPath, 'packages', 'package-two');
  await snapshotDistFiles(pkgPath);

  expect(unsafeRequire(pkgPath).default).toBe(2);
});

test('needs @babel/runtime', async () => {
  let tmpPath = f.copy('use-babel-runtime');
  await install(tmpPath);

  try {
    await build(tmpPath);
  } catch (err) {
    expect(err).toBeInstanceOf(FatalError);
    expect(err.message).toMatchInlineSnapshot(
      `"Babel helpers (functions inserted by Babel transforms) should be imported from the @babel/runtime package to reduce bundle size but @babel/runtime is not in the dependencies of use-babel-runtime, please add it there."`
    );
    return;
  }
  expect(true).toBe(false);
});
