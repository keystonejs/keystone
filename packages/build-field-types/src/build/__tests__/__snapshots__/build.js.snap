// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@babel/runtime installed: babel-runtime-installed.cjs.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

class Bar {}

class Foo extends Bar {}

exports.Foo = Foo;
"
`;

exports[`@babel/runtime installed: babel-runtime-installed.esm.js 1`] = `
"class Bar {}

class Foo extends Bar {}

export { Foo };
"
`;

exports[`monorepo single package: package-two-single-package.cjs.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var index = 2;
exports.default = index;
"
`;

exports[`monorepo single package: package-two-single-package.esm.js 1`] = `
"var index = 2;
export default index;
"
`;

exports[`monorepo: package-one.cjs.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var index = 1;
exports.default = index;
"
`;

exports[`monorepo: package-one.esm.js 1`] = `
"var index = 1;
export default index;
"
`;

exports[`monorepo: package-two.cjs.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var index = 2;
exports.default = index;
"
`;

exports[`monorepo: package-two.esm.js 1`] = `
"var index = 2;
export default index;
"
`;
